### ISSUE TITLE
Cannot directly load files from GSEA

### ISSUE THREAD

**Issue by mikekucera** [BaderLab/EnrichmentMapApp#357]
Bug originally reported in the Cytoscape Jira: 
https://cytoscape.atlassian.net/browse/CSD-21

> Hi, I've finished running GSEA with my RNAseq data. As I tried to use the function Enrichment Map Visualization, it launch Cytoscape but doesn't load the data I just loaded and analyzed.
> Cytoscape version: 3.7.0
> GSEA version: 3.0
> OS: Windows

It sounds like the function in GSEA that opens Cytoscape automatically is not working.

**Comment by risserlin** [comment]
How do I respond to this bug on the https://cytoscape.atlassian.net website?  There is an issue GSEA/EM integration with Cytoscape 3.7.

**Comment by risserlin** [comment]
looks like with cytoscape 3.7 the first time you press "Build enrichment map" from GSEA it gives an error "cytoscape has closed, please re-open it" but cytoscape is still open.  
If you press the "Build enrichment map" button again it will build the network no problem. and every subsequent network build no problem. 

Looks like the initial connection to cytoscape is not working but it is able to repair it on subsequent attempts.  

Alternately you can uninstall cytoscape 3.7 and then go into GSEA preferences and change the cytoscape version to 3.6.1.  shutdown gsea and restart for the settting to take effect.  There should be no issues using cytoscape 3.6.1

**Comment by dotasek** [comment]
@risserlin see my response here: https://cytoscape.atlassian.net/browse/CYTOSCAPE-12477

**Comment by mikekucera** [comment]
I have GSEA running in the eclipse debugger. I'll try to see if I can find the source of the problem on GSEA's end.

**Comment by risserlin** [comment]
@mikekucera I don't think it is a GSEA issue.  I think it might be what @dotasek suggested.  I will try it out and see if it is related to the cyrest re-start.

**Comment by mikekucera** [comment]
I found another related bug in gsea. If you have 3.6.1 and 3.7.0 installed
it always launches 3.6.1 even if you specify 3.7.0 in the prefernences.

**Comment by dotasek** [comment]
> I found another related bug in gsea. If you have 3.6.1 and 3.7.0 installed it always launches 3.6.1 even if you specify 3.7.0 in the prefernences.

This would make the issue worse, because each time Cytoscape is switched, it would check the last installed CyREST, discover that it changed and request a reboot.

**Comment by mikekucera** [comment]
Here's what I've found...

1) There is a bug in GSEA when it compares version numbers. It thinks 3.6.1 is a higher version than 3.7.0. So if you have both installed it will always launch 3.6.1 even if 3.7.0 is already running. Setting the version to 3.7.0 in the preferences panel has no effect. I have to uninstall 3.6.1 for it to recognize 3.7.0. (Note, if 3.7.1 is installed then it will probably pick that.)

Ruth, I can fix this bug myself and send them a pull request. How difficult is it to get them to release a new version of GSEA with the fix? Should we bother trying to fix it?

2) If you load GSEA data without an edb folder EnrichmentMap will throw an exception, you get an error message in GSEA.

I haven't touched the gsea command, so this has always been an issue I guess. The original bug on jira is very short on details. I'm going to ask if this is what they're getting.

3) When 3.7.0 is launched for the first time and you see the CyREST popup GSEA can't connect to it. GSEA says "Cytoscape appears to have been closed. Just re-launcing it...". I have to shutdown and restart Cytoscape. After that it works fine.

David, is there anything we can do about this? I'm guessing probably not.

Note, I'm not getting the behaviour Ruth describes above. I have to restart cytoscape to get it to work. Just clicking the build button again doesn't work.

**Comment by risserlin** [comment]
@mikekucera That is weird. I have both 3.7 and 3.6.1 installed and it launches 3.7 on my machine.

You can't load an enrichment result from GSEA without the edb directory.  It is why we enabled the edb support.

**Comment by mikekucera** [comment]
Actually I see how that could happen...

The bug is here:
https://github.com/GSEA-MSigDB/gsea-desktop/blob/60835bf6ff0869f2072b30e0551c170dc9b319e2/src/edu/mit/broad/cytoscape/CytoscapeLocationSingleton.java#L171

The expression in the if statement is totally wrong. It compares the major, minor, and point parts of the version number on equal footing. The result depends on the order of the cyto_list array. If 3.7.0 comes after 3.6.1 in the array it should pick 3.7.0, but if 3.7.0 comes before 3.6.1 then it will pick 3.6.1. The list of versions comes from the java.io.File.list() method, and the javadoc makes no mention of how the returned array is ordered, so its probably platform dependent.

**Comment by mikekucera** [comment]
So in conclusion...

1. The version bug is minor. It only happens in certain situations, and on certain platforms, and there's an easy workaround (uninstall 3.6.1).
2. The edb folder is required, this is not a bug.
3. There's nothing we can do about the CyREST popup in 3.7.1.

So what can we do? Just make mention of the CyREST issue in the EM docs?

**Comment by risserlin** [comment]
Just make mention of the cyrest issue in the EM docs.  And people might find this thread as well.  It is fixable.

There are a lot of issue with how the cytoscape version is calculated.  I don't think it ever worked well for linux/ubuntu either but when we added the GSEA support cyrest wasn't even available.  We launched cytoscape with a -S option and a command script.   When I updated it to support cyrest I should have changed how it calculated the version then using the cyrest version command but I wanted to make the change minimal.
